compiler/rustc_trait_selection/src/solve/assembly.rs

```diff
@@ -364,8 +364,9 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
             | ty::Tuple(_)
             | ty::Param(_)
             | ty::Placeholder(..)
-            | ty::Infer(_)
+            | ty::Infer(ty::IntVar(_) | ty::FloatVar(_))
             | ty::Error(_) => return,
+            ty::Infer(_) => bug!(),
             ty::Bound(..) => bug!("unexpected bound type: {goal:?}"),
```

Comment:
nit: consistentency for `bug` statements. I guess always using `bug!("unexpected self type for ...")` is probably the most straightforward?